bin/lint.sh

```diff
@@ -1,23 +1,88 @@
-#!/bin/bash
+#!/bin/sh
 
-set -euo pipefail
+set -e
 
 POETRY_RUN="poetry run"
-
 CURRENT_DIR=$(CDPATH= cd -- "$(dirname -- "$0")" && pwd)
 BASE_DIR="$(dirname "$CURRENT_DIR")"
 HAS_GIT="$(command -v git || echo '')"
 echo $HAS_GIT
 
-$POETRY_RUN bandit -lll --recursive "${BASE_DIR}" --exclude "${BASE_DIR}/poetry.lock,${BASE_DIR}/.venv,${BASE_DIR}/.mypy,${BASE_DIR}/build"
+bandit () {
+  $POETRY_RUN bandit -lll --recursive "${BASE_DIR}" --exclude "${BASE_DIR}/poetry.lock,${BASE_DIR}/.venv,${BASE_DIR}/.mypy,${BASE_DIR}/build"
+}
+black () {
+  $POETRY_RUN black ${check:+--check} "${BASE_DIR}"
+}
+detect_secrets () {
+  if [ -n "$HAS_GIT" ]; then
+    # Scan only files fixed into the repo, omit poetry.lock
+    FILES_TO_SCAN=`git ls-tree --full-tree -r --name-only HEAD | grep -v poetry.lock`
+    $POETRY_RUN detect-secrets-hook $FILES_TO_SCAN --baseline .secrets.baseline
+  else
+    echo "Git not available. Skip detect-secrets"
+  fi
+}
+isort () {
+  $POETRY_RUN isort ${check:+--check-only} "${BASE_DIR}"
+}
+pylint () {
+  $POETRY_RUN pylint "${BASE_DIR}/ctms" "${BASE_DIR}/tests/unit"
+}
+mypy () {
+  $POETRY_RUN mypy "${BASE_DIR}/ctms"
+}
+all () {
+  echo "running black"
+  black
+  echo "running isort"
+  isort
+  echo "running mypy"
+  mypy
+  echo "running pylint"
+  pylint
+  echo "running bandit"
+  bandit
+  echo "running detect_secrets"
+  detect_secrets
+}
 
-if [ -n "$HAS_GIT" ]; then
-    # Scan only files checked into the repo, omit poetry.lock
-    SECRETS_TO_SCAN=`git ls-tree --full-tree -r --name-only HEAD | grep -v poetry.lock`
-    $POETRY_RUN detect-secrets-hook $SECRETS_TO_SCAN --baseline .secrets.baseline
-fi
+usage () {
+  echo "Usage: bin/lint.sh [OPTION]"
+  echo " run linting checks"
+  echo "Options":
+  echo "  bandit"
+  echo "  black [--fix]"
+  echo "  detect-secrets"
+  echo "  isort [--fix]"
+  echo "  mypy"
+  echo "  pylint"
+  echo "  yamllint"
+}
 
-$POETRY_RUN isort --check-only "${BASE_DIR}"
-$POETRY_RUN black --check "${BASE_DIR}"
-$POETRY_RUN mypy "${BASE_DIR}/ctms"
-$POETRY_RUN pylint "${BASE_DIR}/ctms" "${BASE_DIR}/tests/unit"
+subcommand='';
+check="true"
+if [ -z $1 ]; then
+  all
+else
+  subcommand=$1; shift
+  case $subcommand in
+    "black" | "isort")
+      case $1 in
+        "--fix")
+          check=""
+        ;;
+      esac
+      case $subcommand in
+        "isort") isort;;
+        "black") black;;
+      esac
+    ;;
```

Comment:
This is one of those cases where "the demo / temporary" fix becomes the "permanent/canonical" fix 🫣. I don't know if this is the "right way" to handle this logic from a shell scripting perspective. For example, there are a [ton of ways](https://stackoverflow.com/questions/3601515/how-to-check-if-a-variable-is-set-in-bash) to test if a variable is set. I don't know if `[ ]` is the right one. I almost wrote the JBI version as a Python script to avoid this headache.

If this passes the sniff test, cool, but if you want to hack on this approach at all, I'm all for it.

Reply:
Honestly, this is fiiiiine. Especially because our interface is `make lint` and `make format`, it's not an interface we offer to the world. Therefore the way we manage empty parameters in this internal shell script is not the most crucial thing. Plus, `if [ -z $1 ]; then` is not wrong, it just doesn't distinguish empty from unset